gluster/block/cmd/glusterblock-provisioner/glusterblock-provisioner.go

```diff
@@ -432,65 +435,72 @@ func parseClassParameters(params map[string]string, kubeclient kubernetes.Interf
+				glog.Errorf("glusterblock: block mode args has to be set if this opmode is set")
+
 			} else {
-				return nil, fmt.Errorf("StorageClass for provisioner %s contains wrong number of arguments for %s", "glusterblock", parseOpmode)
+				parseOpmodeInfo := dstrings.Split(blkmodeArgs, "=")
+				if len(parseOpmodeInfo) >= 2 {
+					argsDict, err := parseBlockModeArgs(cfg.opMode, blkmodeArgs)
+					if err != nil {
+						glog.Errorf("Failed to parse gluster-block arguments")
+						return nil, fmt.Errorf("Failed to parse gluster-block arguments")
+					}
+					cfg.blockModeArgs = *argsDict
+				} else {
+					return nil, fmt.Errorf("StorageClass for provisioner %s contains wrong number of arguments for %s", "glusterblock", parseOpmode)
+				}
 			}
 
 			// Heketi Opmode
 		case "heketi":
-			if len(parseOpmodeInfo) >= 2 {
-				cfg.opMode = "heketi"
-				// TODO:
-			} else {
-				return nil, fmt.Errorf("StorageClass for provisioner %s contains wrong number of arguments for %s", "heketi", parseOpmode)
+			cfg.opMode = "heketi"
+			err := parseHeketiModeArgs(&cfg)
+			if err != nil {
+				glog.Errorf("Failed to parse gluster-block arguments")
+				return nil, fmt.Errorf("Failed to parse gluster-block arguments")
 			}
-
 		default:
-			return nil, fmt.Errorf("StorageClass for provisioner %s contains unknown [%v] parameter", "glusterblock", parseOpmodeInfo[0])
+			return nil, fmt.Errorf("StorageClass for provisioner %s contains unknown [%v] parameter", "glusterblock", parseOpmode)
 		}
 	}
 
-	if !authEnabled {
-		cfg.user = ""
-		cfg.secretName = ""
-		cfg.secretNamespace = ""
-		cfg.userKey = ""
-		cfg.secretValue = ""
+	if len(cfg.url) == 0 && cfg.opMode == "heketi" {
+		return nil, fmt.Errorf("StorageClass for provisioner %s must contain 'resturl' parameter", "glusterblock")
 	}
 
-	if len(cfg.secretName) != 0 || len(cfg.secretNamespace) != 0 {
-		// secretName + Namespace has precedence over userKey
-		if len(cfg.secretName) != 0 && len(cfg.secretNamespace) != 0 {
-			cfg.secretValue, err = parseSecret(cfg.secretNamespace, cfg.secretName, kubeclient)
-			if err != nil {
-				return nil, err
+	if cfg.opMode == "heketi" {
+		if !authEnabled {
+			cfg.user = ""
+			cfg.secretName = ""
+			cfg.secretNamespace = ""
+			cfg.userKey = ""
+			cfg.secretValue = ""
+		}
+
+		if len(cfg.secretName) != 0 || len(cfg.secretNamespace) != 0 {
+			// secretName + Namespace has precedence over userKey
+			if len(cfg.secretName) != 0 && len(cfg.secretNamespace) != 0 {
+				cfg.secretValue, err = parseSecret(cfg.secretNamespace, cfg.secretName, kubeclient)
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				return nil, fmt.Errorf("StorageClass for provisioner %q must have secretNamespace and secretName either both set or both empty", "glusterblock")
+
 			}
 		} else {
-			return nil, fmt.Errorf("StorageClass for provisioner %q must have secretNamespace and secretName either both set or both empty", "glusterblock")
+			cfg.secretValue = cfg.userKey
```

Comment:
what if userKey is blank?

IMO it would be clearer if it did all the validation of params beforehand in a different function instead of waiting until they are needed.